proof/proof.go

```diff
@@ -0,0 +1,160 @@
+	}
+
+	// Now append the path from this node to the root of size2.
+	p := nodes(index, level, size2)
+	p.IDs = append(proof, p.IDs...)
+	if len(proof) == 1 && p.begin < p.end {
+		p.begin++
+		p.end++
+	}
+	return p, nil
+}
+
+// nodes returns the node IDs necessary to prove that the (level, index) node
+// is included in the Merkle tree of the given size.
+func nodes(index uint64, level uint, size uint64) Nodes {
+	// [begin, end) is the leaves range covered by the (level, index) node.
+	begin, end := index<<level, (index+1)<<level
+	// To prove inclusion of range [begin, end), we only need nodes of compact
+	// range [0, begin) and [end, size). Further down, we need the nodes ordered
+	// by level from leaves towards the root.
+	left := reverse(compact.RangeNodes(0, begin))
+	// We decompose the [end, size) range into [end, end+l) and [end+l, size).
+	// The first one (named `middle` here) contains all the nodes that don't have
+	// a left sibling within [end, size), and the second one (named `right`
+	// below) contains all the nodes that don't have a right sibling.
+	l, _ := compact.Decompose(end, size)
+	middle := compact.RangeNodes(end, end+l)
+	// Nodes that don't have a right sibling (i.e. the right border of the tree)
+	// are special, because their hashes are collapsed into a single "ephemeral"
+	// hash. It can be derived from the hashes of compact range [end+l, size).
+	right := reverse(compact.RangeNodes(end+l, size))
+
+	// The level in the ordered list of nodes where the rehashed nodes appear in
+	// lieu of the "ephemeral" node. This is equal to the level where the path to
+	// the `begin` index diverges from the path to `size`.
+	rehashLevel := uint(bits.Len64(begin^size) - 1)
+	var rehashBegin, rehashEnd int
+
+	// Merge the three compact ranges into a single proof ordered by node level
+	// from leaves towards the root, i.e. the format specified in RFC 6962.
+	proof := make([]compact.NodeID, 0, len(left)+len(middle)+len(right))
+	i, j := 0, 0
+	for l, levels := level, uint(bits.Len64(size-1)); l < levels; l++ {
+		if i < len(left) && left[i].Level == l {
+			proof = append(proof, left[i])
+			i++
+		} else if j < len(middle) && middle[j].Level == l {
+			proof = append(proof, middle[j])
+			j++
+		}
+		if l == rehashLevel {
+			proof = append(proof, right...)
+			if len(right) > 1 {
+				rehashBegin = len(proof) - len(right)
+				rehashEnd = len(proof)
+			}
+		}
+	}
+
+	return Nodes{IDs: proof, begin: rehashBegin, end: rehashEnd}
+}
+
+// Rehash computes the proof based on the slice of node hashes corresponding to
+// their IDs in the n.IDs field. The slices must be of the same length. The hc
+// parameter computes a node's hash based on hashes of its children.
+//
+// Warning: The passed-in slice of hashes can be modified in-place.
+func (n Nodes) Rehash(h [][]byte, hc func(left, right []byte) []byte) ([][]byte, error) {
+	if len(h) != len(n.IDs) {
+		return nil, errors.New("slice lengths mismatch")
```

Comment:
Maybe be a bit more explicit in the error here to help debugging? (e.g. `fmt.Errorf("got %d hashes but expected %d", len(h), len(n.IDs)` or something along those lines?)